### PR TITLE
Feature/register

### PR DESCRIPTION
**Pull request by spuder**
This is a working addition to register tentacles

Works on both listening & polling nodes. It is documented in the readme and lists known limitations. 
Is properly idempotent by querying the octopus api using powershell and the octopus dll. 

It is robust enough that I think most people should be able to use it as is.

## Known landmines for discussion

- Only tested on Windows 2012R2 with Chef client 12.12.x. 

- Listening tentacles work, but have limited testing because my local octopus deploy can't find the vm on the other side of the virtualbox (test kitchen) NAT. Polling tentacles are well tested.

- Could not get rubocop to work with the linked gist. Gist returns 404, removed from .rubocop.yml

- Adds a dependency on windows_firewall cookbook. 

- [x] Polling & Listening tentacles default to different ports (10933 vs 10943). Adds a new attribute `polling_port` with sane default. **Should `port` be renamed to `listening_port` for consistency?** 

- [x] If the octopus server firewall is blocking inbound 10933 or 10943, the guard interpreter will skip the tentacle registration. **Is this the desired behavior or should the entire chef run fail if tentacle registration fails?** 

- No tests since tentacle registration needs an API token. [While it is possible to generate an api key with powershell, it isn't trivial.](http://help.octopusdeploy.com/discussions/questions/3553-generate-api-key-from-command-line)

### CODE REVIEW

File: resources/tentacle.rb [hunk 1]
```diff
@@ -18,7 +18,7 @@
 # limitations under the License.
 #
 
-actions :install, :configure, :remove, :uninstall
+actions :install, :configure, :register, :remove, :uninstall
```
Comment (brentm5) [inline resources/tentacle.rb]: Since we are adding the register action I think we should make the version not required so it doesn't need to be passed in.  Then validation should be added for the other methods to keep the same behavior

```
attribute :version, kind_of: String, required: true
```

File: providers/tentacle.rb [hunk 2]
```diff
@@ -129,6 +129,83 @@
...
+    cwd tentacle_install_location
+    code <<-EOH
+      .\\Tentacle.exe register-with --instance "#{instance}" --server "#{server}" --name "#{node.name}" --apiKey "#{api_key}" --comms-style "TentacleActive" --server-comms-port "#{polling_port}" --force --environment "#{environment}" #{role_expression} --console
+      #{catch_powershell_error('Machine registered successfully')}
+    EOH
+    only_if { polling == 'True' }
+    guard_interpreter :powershell_script
+    convert_boolean_return true
+    only_if <<-EOH
+      Add-Type -Path 'C:\\Program Files\\Octopus Deploy\\Tentacle\\Newtonsoft.Json.dll'
```
Comment (brentm5) [inline providers/tentacle.rb]: I think you should do the following just to make it a little less rigid

```
cd #{tentacle_install_location}
Add-Type -Path 'Newtonsoft.Json.dll'
```

File: providers/tentacle.rb [hunk 3]
```diff
@@ -129,6 +129,83 @@
...
+  end
+
+  # # Itterate over every role and make a big long string like:
+  # # --role "web" --role "database" --role "app-server"
+  role_expression = ''
+  roles.each do |a_role|
+    role_expression << "--role \"#{a_role}\" "
+  end
+
+  register_listening_instance = powershell_script "register-listening-tentacle-instance-#{instance}" do
```
Comment (brentm5) [inline providers/tentacle.rb]: I think we can combine these two blocks to be the same and just abstract out some of the logic to generate the command line arguments. Something like this 

```ruby
def communication(polling, port)
  if polling
    " --comms-style \"TentacleActive\" --server-comms-port \"#{port}\" --force"
  else
    "--comms-style TentaclePassive"
  end
end

File: providers/tentacle.rb [hunk 4]
```diff
@@ -129,6 +129,83 @@
...
+    only_if { polling == 'True' }
+    guard_interpreter :powershell_script
+    convert_boolean_return true
+    only_if <<-EOH
+      Add-Type -Path 'C:\\Program Files\\Octopus Deploy\\Tentacle\\Newtonsoft.Json.dll'
+      Add-Type -Path 'C:\\Program Files\\Octopus Deploy\\Tentacle\\Octopus.Client.dll'
+      $endpoint = new-object Octopus.Client.OctopusServerEndpoint '#{server}', '#{api_key}'
+      $repository = new-object Octopus.Client.OctopusRepository $endpoint
+      $tentacle = New-Object Octopus.Client.Model.MachineResource
+      $thumbprint = Invoke-Expression "& 'C:\\Program Files\\Octopus Deploy\\Tentacle\\Tentacle.exe' show-thumbprint --nologo --console"
```
Comment (brentm5) [inline providers/tentacle.rb]: I think for this you can just do this 

```
$thumprint = (& .\\Tentacle.exe show-thumbprint --nologo --console)[1]
```

File: providers/tentacle.rb [hunk 5]
```diff
@@ -129,6 +129,83 @@
...
+  api_key = new_resource.api_key
+  roles = new_resource.roles
+  environment = new_resource.environment
+
+  verify_server(server)
+  verify_api_key(api_key)
+  verify_roles(roles)
+  verify_environment(environment)
+
+  firewall = windows_firewall_rule 'Octopus Deploy Tentacle' do
```
Comment (brentm5) [inline providers/tentacle.rb]: I think this should be a flag or part of configure.  Can we possibly remove this and add it to another PR which is part of the configure?

File: resources/tentacle.rb [hunk 6]
```diff
@@ -29,6 +29,11 @@
 attribute :app_path, kind_of: String, default: 'C:\Octopus\Applications'
 attribute :trusted_cert, kind_of: String
 attribute :port, kind_of: Fixnum, default: 10_933
+attribute :polling_port, kind_of: Fixnum, default: 10_943
```
Comment (brentm5) [inline resources/tentacle.rb]: I think we should just use the same port value for everything.  If its so we can have an intelligent default we can lazy evaluate the default and make it look at `polling` or not

File: README.md [hunk 7]
```diff
@@ -75,6 +80,38 @@ end
...
+Register Listening Tentacle
+# You will first need to generate an api key
+# In Octopus Deploy Server GUI click your Name -> Profile -> API keys
+octopus_deploy_tentacle 'Tentacle' do
+  action :register
+  server 'https://octopus.example.com'
+  api_key '12345678910'
+  roles ['database']
+  environment 'prod'
+  polling false
```
Comment (brentm5) [inline README.md]: I think these can be left out for listening tentacles

Comment (spuder) [inline README.md]: Which can be left out? The role and environment? The server and api key are required for the idempotency check.

File: resources/tentacle.rb (hunk 1 above)
Comment (spuder) [inline resources/tentacle.rb]: Done in a01581ed64acc1f557dbb19c13a2a4005a378684
All new resource attributes have verify methods. Which other methods were needed?

File: README.md (hunk 7 above)
Comment (brentm5) [inline README.md]: For listening tentacles we do not need the polling boolean or the port specified.  So for this example we can actually just leave those attributes off.  We only need to add the values if its  not the defaults.

File: resources/tentacle.rb (hunk 1 above)
Comment (brentm5) [inline resources/tentacle.rb]: Awesome. Ignore my comment about the verify methods.  I forgot this was already done for other resources.  Looks good.

File: providers/tentacle.rb (hunk 4 above)
Comment (spuder) [inline providers/tentacle.rb]: ✅  Done fb1689c0abd7497a9a1f2fe4edf385bbaddfbfdd

File: providers/tentacle.rb (hunk 2 above)
Comment (spuder) [inline providers/tentacle.rb]: ✅  Done fb1689c0abd7497a9a1f2fe4edf385bbaddfbfdd

File: resources/tentacle.rb (hunk 6 above)
Comment (spuder) [inline resources/tentacle.rb]: Lazy evaluation never seems to do what I expect it to. 

This appears to work, is this what you are thinking? 

```ruby
attribute :port, kind_of: Fixnum, default: 10_933 if lazy { fancy_bool(:polling) == "False" }
attribute :port, kind_of: Fixnum, default: 10_943 if lazy { fancy_bool(:polling) == "True" }
```

File: README.md (hunk 7 above)
Comment (spuder) [inline README.md]: ✅  Done

File: resources/tentacle.rb (hunk 6 above)
Comment (brentm5) [inline resources/tentacle.rb]: ummm not sure.  I also think there are issues with lazy evaluation on earlier versions of chef-client.  Maybe the best thing to do is to set the value to nil and if its nil just default it in the resource?

Comment (spuder) [inline resources/tentacle.rb]: ✅  See dbf88193b36812a53a92c716780033ba514e14a2

File: providers/tentacle.rb (hunk 5 above)
Comment (spuder) [inline providers/tentacle.rb]: I don't understand, could you elaborate? 

Are you saying that you want the :configure and the :register resource actions combined into 1? 
That would require that users generate an api key to do configuration which would be a breaking change and probably mandate a major release. By keeping the resource actions separate it maintains backwards compatibility.

File: providers/tentacle.rb [hunk 8]
```diff
@@ -281,3 +280,17 @@ def verify_roles(roles)
 def verify_environment(environment)
   raise 'Environment is required in order to register a Tentacle with Octopus Deploy Server' unless environment
 end
+
+def verify_port(port,polling)
```
Comment (brentm5) [inline providers/tentacle.rb]: I would call this resolve_port and put it in the tentacle library file.  Also I would pass in actual true or false value and just use the fancy_boolean for powershell formatting only.

```
def resolve_port(polling, port)
  port || default_port(polling)
end

def default_port(polling) 
  if polling
    10_943
  else
    10_933
  end
end 
```

File: providers/tentacle.rb [hunk 9]
```diff
@@ -63,8 +63,8 @@
   config_path = new_resource.config_path
   app_path = new_resource.app_path
   trusted_cert = new_resource.trusted_cert
-  port = new_resource.port
   polling = fancy_bool(new_resource.polling)
```
Comment (brentm5) [inline providers/tentacle.rb]: I would only use fancy_bool in powershell scripts.  you can use that function with the `#{fancy_bool(polling)}` I am pretty sure

File: Berksfile [hunk 10]
```diff
@@ -5,3 +5,5 @@ metadata
 group :test do
   cookbook 'octopus-deploy-test', path: './test/cookbooks/octopus-deploy-test'
 end
+
+cookbook 'windows_firewall'
```
Comment (brentm5) [inline Berksfile]: This shouldnt be needed, it will pick it up from the metadata file.  Only test dependencies need to be pushed here

File: metadata.rb [hunk 11]
```diff
@@ -9,6 +9,7 @@
 version '0.5.2'
 
 depends 'windows', '~> 1.38'
+depends 'windows_firewall', '~> 3.0.2'
```
Comment (brentm5) [inline metadata.rb]: Can we move this into another PR and include it on the tentacle configure action.  Ideally this needs to be setup when the tentacle is configured not when we want to register

File: providers/tentacle.rb (hunk 5 above)
Comment (brentm5) [inline providers/tentacle.rb]: No so configuring a tentacle is a requirement to add it to the octo server.  I am simply saying that we should setup the firewall items in the configure section instead of in the register section.  In addition it should also be behind a flag.

File: providers/tentacle.rb (hunk 8 above)
Comment (spuder) [inline providers/tentacle.rb]: Will do, this is much better. 

The reason for using fancy_boolean is so that it is consistent with the :configure resource. 

```
action :configure do
  new_resource = @new_resource
  name = new_resource.name
  instance = new_resource.instance
  checksum = new_resource.checksum
  version = new_resource.version
  upgrades_enabled = new_resource.upgrades_enabled
  home_path = new_resource.home_path
  config_path = new_resource.config_path
  app_path = new_resource.app_path
  trusted_cert = new_resource.trusted_cert
  polling = fancy_bool(new_resource.polling)
```

Should we modify :configure and :register so that it converts boolean to fancy_boolean as late as possible? 

So this: 
      .\\Tentacle.exe configure --instance="#{instance}" --home="#{home_path}" --app="#{app_path}" --port="#{port}" --noListen="#{polling}" --console

Would become

      .\\Tentacle.exe configure --instance="#{instance}" --home="#{home_path}" --app="#{app_path}" --port="#{port}" --noListen="#{fancy_boolean(polling)}" --console

Comment (brentm5) [inline providers/tentacle.rb]: I can make another issue to follow up and do that.  No need to here!

Comment (brentm5) [inline providers/tentacle.rb]: Added #34 to account for this

File: metadata.rb (hunk 11 above)
Comment (spuder) [inline metadata.rb]: ✅ Made #35

File: providers/tentacle.rb (hunk 3 above)
Comment (spuder) [inline providers/tentacle.rb]: I'm going to defer the abstraction of the --comms style since its adding complexity and I can't quite get it working right.

Comment (brentm5) [inline providers/tentacle.rb]: Added #36

File: providers/tentacle.rb [hunk 12]
```diff
@@ -129,6 +129,73 @@
...
+    EOH
+  end
+
+  register_polling_instance = powershell_script "register-polling-tentacle-instance-#{instance}" do
+    cwd tentacle_install_location
+    code <<-EOH
+      .\\Tentacle.exe register-with --instance "#{instance}" --server "#{server}" --name "#{node.name}" --apiKey "#{api_key}" --comms-style "TentacleActive" --server-comms-port "#{port}" --force --environment "#{environment}" #{role_expression} --console
+      #{catch_powershell_error('Registering Tentacle')}
+    EOH
+    only_if { polling == true }
```
Comment (brentm5) [inline providers/tentacle.rb]:
```ruby
only_if { polling }

# not sure if this will work but this would be better

only_if polling
```

File: providers/tentacle.rb [hunk 13]
```diff
@@ -129,6 +129,73 @@
...
+    guard_interpreter :powershell_script
```
Comment (brentm5) [inline providers/tentacle.rb]: I believe you can remove this

File: providers/tentacle.rb [hunk 14]
```diff
@@ -129,6 +129,73 @@
...
+
+  register_listening_instance = powershell_script "register-listening-tentacle-instance-#{instance}" do
+    cwd tentacle_install_location
+    code <<-EOH
+      .\\Tentacle.exe register-with --instance "#{instance}" --server "#{server}" --name "#{node.name}" --apiKey="#{api_key}" #{role_expression} --environment "#{environment}" --comms-style TentaclePassive --console
+      #{catch_powershell_error('Registering Tentacle')}
+    EOH
+    only_if { polling == false }
+    guard_interpreter :powershell_script
+    convert_boolean_return true
```
Comment (brentm5) [inline providers/tentacle.rb]: Also I think this can be removed because its not used anywehre else and we throw an exception.

File: providers/tentacle.rb (hunk 12 above)
Comment (spuder) [inline providers/tentacle.rb]: ✅ Done in fafc2f4e3898693ad96692129cfea80c2eff098e
I'm not positive if the `new_resource.updated_by_last_action` is correct, but it appears to work. 

https://github.com/cvent/octopus-deploy-cookbook/pull/33/commits/fafc2f4e3898693ad96692129cfea80c2eff098e#diff-e771d6166058c16c112c0d2c61321094R173

https://github.com/cvent/octopus-deploy-cookbook/pull/33/commits/fafc2f4e3898693ad96692129cfea80c2eff098e#diff-e771d6166058c16c112c0d2c61321094R193

**Comment by spuder** [comment]
I've squashed all commits and removed the WIP from the title. 

I've tested manually again and it feels pretty bullet proof. 

Let me know if there is anything else you want changed. 

# nitpick

The only part I'm not thrilled about is the powershell script used for idempotency. As written if the octopus deploy server is unavailable for any reason, the recipe will act exactly as if the node is already registered and will silently continue. That may surprise some users. 

I could add another powershell_script resource that tests if the octopus api is available and print a warning if it isn't available. My only hesitation is that it will double the api calls per tentacle per chef run and I don't know how well the octopus deploy api scales. Could it handle 1000 nodes doing 2 api calls ever 30 minutes?

**Comment by brentm5** [comment]
👍  Looks good to me! awesome work.
